src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs

```diff
@@ -94,55 +98,27 @@ internal partial class NavigationBarController
     {
-        // Switch to the UI so we can determine where the user is and determine the state the last time we updated
-        // the UI.
-        await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken).NoThrowAwaitable();
-
-        // Cancellation exceptions are ignored in AsyncBatchingWorkQueue, so return without throwing if cancellation
-        // occurred while switching to the main thread.
-        if (cancellationToken.IsCancellationRequested)
-            return;
+        var lastCaretPosition = positions.Last();
 
-        await SelectItemWorkerAsync(cancellationToken).ConfigureAwait(true);
+        // Can grab this directly here as only this queue ever reads or writes to it.
+        var lastPresentedInfo = _lastPresentedInfo;
 
-        // Once we've computed and selected the latest navbar items, pause ourselves if we're no longer visible.
-        // That way we don't consume any machine resources that the user won't even notice.
-        if (_visibilityTracker?.IsVisible(_subjectBuffer) is false)
-            Pause();
-    }
+        // Make a task that waits indefinitely, or until the cancellation token is signaled.
+        var cancellationTriggeredTask = Task.Delay(-1, cancellationToken);
 
-    private async ValueTask SelectItemWorkerAsync(CancellationToken cancellationToken)
-    {
-        _threadingContext.ThrowIfNotOnUIThread();
+        // Get the task representing the computation of the model.
+        var modelTask = _computeModelQueue.WaitUntilCurrentBatchCompletesAsync();
 
-        var currentView = _presenter.TryGetCurrentView();
-        var caretPosition = currentView?.GetCaretPoint(_subjectBuffer);
-        if (!caretPosition.HasValue)
+        var completedTask = await Task.WhenAny(cancellationTriggeredTask, modelTask).ConfigureAwait(false);
+        if (completedTask == cancellationTriggeredTask)
             return;
 
-        var position = caretPosition.Value.Position;
-        var lastPresentedInfo = _lastPresentedInfo;
-
-        // Jump back to the BG to do any expensive work walking the entire model
-        await TaskScheduler.Default;
-
-        // Ensure the latest model is computed.
-        var model = await _computeModelQueue.WaitUntilCurrentBatchCompletesAsync().ConfigureAwait(true);
-
-        var currentSelectedItem = ComputeSelectedTypeAndMember(model, position, cancellationToken);
+        var model = await modelTask.ConfigureAwait(false);
+        var currentSelectedItem = ComputeSelectedTypeAndMember(model, lastCaretPosition, cancellationToken);
 
-        GetProjectItems(out var projectItems, out var selectedProjectItem);
+        var (projectItems, selectedProjectItem) = GetProjectItems();
         if (Equals(model, lastPresentedInfo.model) &&
             Equals(currentSelectedItem, lastPresentedInfo.selectedInfo) &&
             Equals(selectedProjectItem, lastPresentedInfo.selectedProjectItem) &&
```

Comment:
in the common case (caret moving in the same member) we will bail out immediately, and not hit hte ui thread.  we only need to go back if we're actually updating the UI thread.